src/Guard/TelegramAuthenticator.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+namespace BoShurik\TelegramBotBundle\Guard;
+
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+use Symfony\Component\Security\Guard\Authenticator\AbstractFormLoginAuthenticator;
+use Symfony\Component\Security\Http\Util\TargetPathTrait;
+
+class TelegramAuthenticator extends AbstractFormLoginAuthenticator
+{
+    use TargetPathTrait;
+
+    /**
+     * @var TelegramLoginValidator
+     */
+    private $validator;
+
+    /**
+     * @var UserLoaderInterface|UserFactoryInterface
+     */
+    private $userProvider;
+
+    /**
+     * @var UrlGeneratorInterface
+     */
+    private $urlGenerator;
+
+    /**
+     * @var string
+     */
+    private $guardRoute;
+
+    /**
+     * @var string|null
+     */
+    private $loginRoute;
+
+    /**
+     * @var string
+     */
+    private $defaultTargetRoute;
+
+    public function __construct(TelegramLoginValidator $validator, UserLoaderInterface $userProvider, UrlGeneratorInterface $urlGenerator, string $guardRoute, string $defaultTargetRoute, string $loginRoute = null)
+    {
+        $this->validator = $validator;
+        $this->userProvider = $userProvider;
+        $this->urlGenerator = $urlGenerator;
+        $this->guardRoute = $guardRoute;
+        $this->loginRoute = $loginRoute;
+        $this->defaultTargetRoute = $defaultTargetRoute;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supports(Request $request)
+    {
+        $route = $request->attributes->get('_route');
+
+        return $route === $this->guardRoute;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    final public function getCredentials(Request $request)
+    {
+        return $request->query->all();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getUser($credentials, UserProviderInterface $userProvider)
+    {
+        $this->validator->validate($credentials);
+
+        $user = $this->userProvider->loadByTelegramId($credentials['id']);
+
+        if (!$user && $this->userProvider instanceof UserFactoryInterface) {
```

Comment:
I think it's better to separate `UserLoaderInterface` and `UserFactoryInterface` into different parameters. The reason is that `UserLoaderInterface` will be probably implemented by a doctrine `UserRepository` class. On the other hand `UserFactoryInterface` will probably needs more dependencies that do not needed by a repository (e.g. `EventDispatcher`, `Mailer` etc)

Reply:
Done